crypto/fipsmodule/kdf/sskdf.c

```diff
@@ -0,0 +1,336 @@
+                                      const uint8_t *secret, size_t secret_len,
+                                      const uint8_t *info, size_t info_len) {
+  if (!ctx || !ctx->data || !out || !counter || !secret) {
+    return 0;
+  }
+
+  sskdf_variant_hmac_ctx *variant_ctx = (sskdf_variant_hmac_ctx *)ctx->data;
+
+  if (!variant_ctx->hmac_ctx) {
+    return 0;
+  }
+
+  uint32_t written;
+
+  if (!HMAC_Init_ex(variant_ctx->hmac_ctx, NULL, 0, NULL, NULL) ||
+      !HMAC_Update(variant_ctx->hmac_ctx, &counter[0], SSKDF_COUNTER_SIZE) ||
+      !HMAC_Update(variant_ctx->hmac_ctx, secret, secret_len) ||
+      !HMAC_Update(variant_ctx->hmac_ctx, info, info_len) ||
+      !HMAC_Final(variant_ctx->hmac_ctx, out, &written) || out_len != written) {
+    return 0;
+  }
+
+  return 1;
+}
+
+
+DEFINE_METHOD_FUNCTION(sskdf_variant, sskdf_variant_digest) {
+  memset(out, 0, sizeof(sskdf_variant));
+  out->output_size = sskdf_variant_digest_output_size;
+  out->compute = sskdf_variant_digest_compute;
+}
+
+
+DEFINE_METHOD_FUNCTION(sskdf_variant, sskdf_variant_hmac) {
+  memset(out, 0, sizeof(sskdf_variant));
+  out->output_size = sskdf_variant_hmac_output_size;
+  out->compute = sskdf_variant_hmac_compute;
+}
+
+static int SSKDF(const sskdf_variant *variant, sskdf_variant_ctx *ctx,
+                 uint8_t *out_key, size_t out_len, const uint8_t *secret,
+                 size_t secret_len, const uint8_t *info, size_t info_len) {
+  int ret = 0;
+
+  if (!ctx || !variant) {
+    return 0;
+  }
+
+  if (!out_key || out_len == 0 || !secret || secret_len == 0) {
+    goto err;
+  }
+
+  // output_size is the length in bytes of output of the SSKDF variant auxilary
+  // function (EVP_DigestFinal or HMAC_Final)
+  size_t output_size = variant->output_size(ctx);
+  if (output_size == 0) {
+    goto err;
+  }
+  assert(output_size <= EVP_MAX_MD_SIZE);
+
+  // NIST.SP.800-56Cr2:
+  // 1. If L > 0, then set reps = ceil(L / H_outputBits); otherwise, output an
+  // error indicator and exit this process without performing the remaining
+  // actions
+  uint64_t n = (out_len + output_size - 1) / output_size;
+
+  // NIST.SP.800-56Cr2:
+  // 2. If reps > (2^32 - 1), then output an error indicator and exit this
+  // process without performing the remaining actions
+  if (n > UINT32_MAX) {
+    goto err;
+  }
+
+  // Validate that the sum of the fixed info lengths won't overflow a size_t
+  // which will very based on 32-bit or 64-bit platforms. We will further clamp
+  // this value below.
+  if (info_len > SIZE_MAX - 4 - secret_len) {
+    goto err;
+  }
+
+  // NIST.SP.800-56Cr2: "4. If counter || Z || FixedInfo is more than
+  // max_H_inputBits bits long, then output an error indicator and exit this
+  // process without performing any of the remaining actions"
+  //
+  // UINT32_MAX is a sufficient max value to satisfy this requirement. See
+  // Section 4.2 Table 1 and 2
+  // https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr2.pdf
```

Comment:
can you say a bit more on how you determine this based on those tables?

Reply:
Let me revisit this with our internal customers first to see what the max derived output length they would reasonably require. It seems suspect that a single invocation of the SSKDF is going to ask for 32 GiB of output key material in one-shot with a given shared secret and info string?

OpenSSL's implementation currently places the following limitations that we might wish to impose to have matching capatability between our implementations:
`secret_len <= 1<<30`, `info_len <= 1<<30`, and also limits the requested derivation length to be `derived_key_len <= 1<<30`.

We could relax this restriction, but that could also result in issues in some hypothetical key agreement scheme where the party using AWS-LC is able to arrived to a derived key with larger inputs, while the party using OpenSSL would fail. Hypothetical of course, but such a scenario could arrise if we choose to differ. Regardless I think the call out here is that the values here, and those by OpenSSL are sufficient enough to satisfy the NIST requirements.

Reply:
Some proof that the OpenSSL limits are sufficient to cap the inputs to satisfy the NIST table 2 and 3:

```C
#include <stddef.h>
#include <stdint.h>

#define OPENSSL_SSKDF_MAX (1 << 30)
#define NIST_MAX_APPROXIMATION (1UL << 60)

int check_overflow(size_t a, size_t b, size_t c) {
  if (a > SIZE_MAX - b || a + b > SIZE_MAX - c) {
    return 0;
  }
  return 1;
}

int main() {
  size_t z = OPENSSL_SSKDF_MAX;
  size_t info_len = OPENSSL_SSKDF_MAX;
  size_t counter = UINT32_MAX;

  if (!check_overflow(z, info_len, counter)) {
    return 1;
  }

  if (z + info_len + counter > NIST_MAX_APPROXIMATION) {
    return 2;
  }

  return 0;
}
```

This program returns 0 if you compile and execute it, thus showing the OpenSSL imposed limits are sufficient to appropriately cap using the smallest max input limit opposed by the tables (max `2^64 - 512` in bits). As the sum of the max secret length, max info length, and max counter will not exceed `2^60` which is the nearest power of 2 smaller then `(2^64 - 512)/8`.